pipe_segment/cli/cli.py

```diff
@@ -0,0 +1,95 @@
+import sys
+import logging
+import argparse
+
+from rich.logging import RichHandler
+
+from .commands.segment import Segment
+
+logger = logging.getLogger(__name__)
+
+
+class CLI:
+    """Generic command-line interface."""
+    HELP_VERBOSE = 'whether to run with DEBUG log level (default: %(default)s).'
+    HELP_LOG_FILE = "file to send logging output to."
+
+    def __init__(self, args):
+        self._args = args
+
+        self._setup_logger()
+        self._add_commands()
+        self._parse_args()
+
+    def execute(self):
+        return self.COMMANDS[self.command].run(self.args, self.extra_args)
+
+    def _add_commands(self):
+        self.parser = argparse.ArgumentParser(
+            prog=self.NAME, description=self.DESCRIPTION, formatter_class=self.formatter())
+
+        self.parser.add_argument('-v', '--verbose', action='store_true', help=self.HELP_VERBOSE)
+        self.parser.add_argument('-l', '--log_file', metavar='\b', help=self.HELP_LOG_FILE)
+
+        self.subparsers = self.parser.add_subparsers(dest='command', help='available commands')
```

Comment:
You can do this this way, but just pointing out an alternative setup I usually use. Instead of having a mapping between command and run function (with `Segment.Name` and so on) you could do this: 

```python
self.parser.add_subparser(
  #...
  action='store_const',
  const=Segment,
)
```

Then, when executing you can just run `self.command.run(...)` as you have the actual command class stored in `args.command` instead of the command name.

Reply:
Hi!
I'm trying to understand, but I think this is not the way subparsers work/should be used.
For starters, `add_subparsers` method does not accept a `const` argument:

```bash
    self.subparsers = self.parser.add_subparsers(action='store_true', const=Segment, help='available commands')
  File "/usr/lib/python3.8/argparse.py", line 1753, in add_subparsers
    action = parsers_class(option_strings=[], **kwargs)
TypeError: __init__() got an unexpected keyword argument 'const'
```

Besides, from [documentation](https://docs.python.org/3/library/argparse.html#sub-commands), you only call once `add_subparsers`, and then you add parsers to it, for each sub-command. 

Maybe you are referring to the [add_argument](https://docs.python.org/3/library/argparse.html#the-add-argument-method) that does accept those arguments?

Reply:
yeah, sorry, I got confused. What I mean is at the end of [the argparse subcommands section](https://docs.python.org/3/library/argparse.html#sub-commands). The approach they recommend is to setup a default on each subparser that sets the run function, so that when the subparser is activated you get the run function set as a param. Something like this:

```python
def bar(args):
  # Do whatever this subcommand does

parser_bar = subparsers.add_parser('bar')
parser_bar.add_argument('z')
parser_bar.set_defaults(run=bar)

# Later in main
args = # parse argv
args.run(args).
```

Reply:
Ok, nice!.

Done.